### PR TITLE
win: fix fstat for pipes and character files

### PR DESCRIPTION
**Pull request by StefanStojanovic**
On Windows, when calling `uv_fs_fstat` for `stdin`, `stdout` or `stderr` file descriptors, an error is thrown, because the call to `NtQueryInformationFile` fails with Nt status code `0xC0000010 (STATUS_INVALID_DEVICE_REQUEST)`. Probably one of the easiest ways to reproduce this would be to run the following command in Node.JS `require('fs').fstatSync(0)`. At the same time, this is a completely regular script to run on Linux as it will fetch stdin stats and throws no errors.

The solution is based on the way Windows Kit implements `fstat` in `CRT` (more details on that can be found in Windows Kit sources eg. `C:\Program Files (x86)\Windows Kits\10\Source\10.0.19041.0\ucrt\filesystem\stat.cpp`). When dealing with regular disk files, there are no changes ie. previously existing logic is used. However, when a character or pipe file is detected, a new logic to fill the response data is triggered.

A new test is added to check this behavior and it's called `fs_fstat_stdio`. It runs on all platforms, although validation on Windows is the most detailed one.

Refs: https://github.com/nodejs/node/issues/40006

### CODE REVIEW

**Review by vtjnash — CHANGES_REQUESTED** [review]
`fd` is not always available, which is why we use the native Win32 APIs rather than the msvcrt compatibility shim--and there is already a uv_guess_handle function for getting this information in libuv

**Comment by StefanStojanovic** [comment]
Implementation is changed to use `uv_guess_handle`. Even though no changes were planned in `fs__stat_handle`, they turned out to be required in order to cover null device usage eg. `node script.js > nul`

**Comment by StefanStojanovic** [comment]
@vtjnash is there something else I should do here to help move this forward?

**Review by vtjnash — COMMENTED** [review]
Sorry for the delay. I had been meaning to look up what reasonable defaults would look like for these structs to satisfy what properties an application might expect to see here

File: src/win/fs.c [hunk 1]
```diff
@@ -1915,6 +1936,37 @@ INLINE static void fs__stat_impl(uv_fs_t* req, int do_lstat) {
...
+  /* Devices and pipes are processed identically. There is no more information
+   * for them from any API. Fields are set as reasonably as possible and the
+   * function returns. */
+  case UV_TTY:
+  case UV_NAMED_PIPE:
+    memset(statbuf, 0, sizeof(uv_stat_t));
+    statbuf->st_mode = file_type == UV_TTY ? _S_IFCHR : _S_IFIFO;
+    statbuf->st_nlink = 1;
+    statbuf->st_rdev = (uint64_t)fd;
+    statbuf->st_dev = (uint64_t)fd;
```
[the remark this answers is not in the document]
Comment (StefanStojanovic) [reply src/win/fs.c]: Unfortunately, that link doesn't show the full list. I've tested with "CONIN$" and "CONOUT$", and they are `FILE_DEVICE_CONSOLE (0x50)`. I've pushed a commit with that change and also adapted tests to your previous changes.

[the remark this answers is not in the document]
Comment (vtjnash) [reply src/win/fs.c]: Awesome, thanks for checking!

File: src/win/fs.c [hunk 2]
```diff
@@ -1915,6 +1942,37 @@ INLINE static void fs__stat_impl(uv_fs_t* req, int do_lstat) {
...
+  /* Devices and pipes are processed identically. There is no more information
+   * for them from any API. Fields are set as reasonably as possible and the
+   * function returns. */
+  case UV_TTY:
+  case UV_NAMED_PIPE:
+    memset(statbuf, 0, sizeof(uv_stat_t));
+    statbuf->st_mode = file_type == UV_TTY ? _S_IFCHR : _S_IFIFO;
+    statbuf->st_nlink = 1;
+    statbuf->st_rdev = (file_type == UV_TTY ? FILE_DEVICE_CONSOLE : FILE_DEVICE_NAMED_PIPE) << 16;
+    statbuf->st_ino = (uint64_t)uv_get_osfhandle(fd);
```
Comment (vtjnash) [inline src/win/fs.c:1964]: minor nit: I missed that this value is already one of the arguments
```suggestion
    statbuf->st_ino = (uint64_t) handle;
```

**Review by vtjnash — APPROVED** [review]

**Pull request merged**